File: src/data/shipsystems/scripts/aEP_NCReloadScript.java
package data.shipsystems.scripts;

import com.fs.starfarer.api.Global;
import com.fs.starfarer.api.combat.CombatEntityAPI;
import com.fs.starfarer.api.combat.MutableShipStatsAPI;
import com.fs.starfarer.api.combat.ShipAPI;
import com.fs.starfarer.api.combat.WeaponAPI;
import com.fs.starfarer.api.impl.combat.BaseShipSystemScript;
import combat.impl.VEs.aEP_MovingSmoke;
import combat.impl.aEP_BaseCombatEffect;
import combat.plugin.aEP_CombatEffectPlugin;
import combat.util.aEP_Tool;
import org.lazywizard.lazylib.MathUtils;
import org.lwjgl.util.vector.Vector2f;

import java.awt.*;
import java.util.ArrayList;
import java.util.List;

public class aEP_NCReloadScript extends BaseShipSystemScript
{

  static final float RELOAD_COOLDOWN_MULT = 0.25f;//0.25 = 75 percent off
  static final float GLOW_INTERVAL = 0.05f;
  static final float GLOW_TIME = 0.5f;
  static final float GLOW_SIZE = 25f;
  static final float GLOW_SIZE_INTERVAL = 30f;
  static List smokeSprites = new ArrayList();
  private ShipAPI ship;


  private boolean didRefresherOrb = false;
  private boolean didOnce = false;
  {
    smokeSprites.add("aEP_FX.thick_smoke01");
    smokeSprites.add("aEP_FX.thick_smoke02");
    smokeSprites.add("aEP_FX.thick_smoke03");
  }

  public void apply(MutableShipStatsAPI stats, String id, State state, float effectLevel) {
    this.ship = (ShipAPI) stats.getEntity();
    if (ship == null) return;


    if (!didRefresherOrb) {
      aEP_CombatEffectPlugin.Mod.addEffect(new RefresherOrb(ship));
      didRefresherOrb = true;
    }

    if (!didOnce) {
      for (WeaponAPI w : ship.getAllWeapons()) {
        if (w.getSpec().getType() == WeaponAPI.WeaponType.MISSILE && w.getCooldownRemaining() > 0.5) {
          w.beginSelectionFlash();
          w.setRemainingCooldownTo(w.getCooldownRemaining() * RELOAD_COOLDOWN_MULT);
          int cloudNum = 0;
          while (cloudNum < 5) {
            //add cloud
            int x = (int) (Math.random() * smokeSprites.size());
            int colorMult = (int) (Math.random() * 40f) + 40;
            float sizeMult = (float) Math.random();
            aEP_MovingSmoke ms = new aEP_MovingSmoke(w.getLocation());
            ms.setInitVel(aEP_Tool.Util.speed2Velocity(MathUtils.getRandomNumberInRange(0f,360f),1.2f));
            ms.setLifeTime(4f);
            ms.setFadeIn(0.1f);
            ms.setFadeOut(0.7f);
            ms.setSize(20 + 40 * sizeMult);
            ms.setColor(new Color(colorMult, colorMult, colorMult, (int) (Math.random() * 40) + 200));
            aEP_CombatEffectPlugin.Mod.addEffect(ms);
            cloudNum = cloudNum + 1;
          }
        }
        didOnce = true;
      }
    }
  }


  public void unapply(MutableShipStatsAPI stats, String id) {
    didRefresherOrb = false;
    didOnce = false;
  }

  public StatusData getStatusData(int index, State state, float effectLevel) {
    return null;
  }

  public static class RefresherOrb extends aEP_BaseCombatEffect
  {
    CombatEntityAPI target;
    float length;
    float advanceDist;
    float glowTimer = GLOW_INTERVAL;

    RefresherOrb(CombatEntityAPI target) {
      this.target = target;
      this.length = -target.getCollisionRadius();
      int num = (int) ((target.getCollisionRadius() * 2f) / GLOW_SIZE_INTERVAL);
      advanceDist = GLOW_SIZE_INTERVAL + ((target.getCollisionRadius() * 2f) % GLOW_SIZE_INTERVAL) / num;
    }

    @Override
    public void advance(float amount) {
      super.advance(amount);
      if (length > target.getCollisionRadius() * 3f) {
        cleanup();
      }

      glowTimer = glowTimer + amount;
      if (glowTimer > GLOW_INTERVAL) {
        glowTimer = glowTimer - GLOW_INTERVAL;
        float xLength = length;
        float yLength = target.getCollisionRadius();
        while (yLength >= -target.getCollisionRadius()) {

          Vector2f toSpawn = new Vector2f(target.getLocation().x + xLength, target.getLocation().y + yLength);
          if (MathUtils.getDistance(toSpawn, target.getLocation()) <= target.getCollisionRadius()) {
            int alpha = (int) ((1 - MathUtils.getDistance(toSpawn, target.getLocation()) / target.getCollisionRadius()) * 250);
            alpha = MathUtils.clamp(alpha, 0, 250);
            Global.getCombatEngine().addSmoothParticle(toSpawn,
                    new Vector2f(0, 0),
                    GLOW_SIZE,
                    1,
                    GLOW_TIME,
                    new Color(0, 250, 0, alpha));
            Global.getCombatEngine().addSmoothParticle(toSpawn,
                    new Vector2f(0, 0),
                    GLOW_SIZE,
                    1,
                    GLOW_TIME,
                    new Color(0, 250, 0, alpha));
          }
          yLength = yLength - advanceDist;
          xLength = xLength - advanceDist;
        }
        length = length + advanceDist;

      }
      super.advanceImpl(amount);
    }


  }
}

File: src/data/shipsystems/scripts/aEP_NCReloadScript.kt
package data.shipsystems.scripts

import com.fs.starfarer.api.Global
import combat.plugin.aEP_CombatEffectPlugin.Mod.addEffect
import combat.util.aEP_Tool.Util.speed2Velocity
import com.fs.starfarer.api.impl.combat.BaseShipSystemScript
import com.fs.starfarer.api.combat.ShipAPI
import com.fs.starfarer.api.combat.MutableShipStatsAPI
import com.fs.starfarer.api.plugins.ShipSystemStatsScript
import combat.plugin.aEP_CombatEffectPlugin
import data.shipsystems.scripts.aEP_NCReloadScript.RefresherOrb
import com.fs.starfarer.api.combat.WeaponAPI
import data.shipsystems.scripts.aEP_NCReloadScript
import combat.impl.VEs.aEP_MovingSmoke
import combat.util.aEP_Tool
import org.lazywizard.lazylib.MathUtils
import com.fs.starfarer.api.plugins.ShipSystemStatsScript.StatusData
import com.fs.starfarer.api.combat.CombatEntityAPI
import combat.impl.aEP_BaseCombatEffect
import combat.util.aEP_ID
import data.hullmods.aEP_MissilePlatform
import org.lwjgl.util.vector.Vector2f
import java.awt.Color
import java.util.ArrayList

class aEP_NCReloadScript : BaseShipSystemScript() {
  var didVisual = false


  companion object {
    const val RELOAD_COOLDOWN_MULT = 0.25f //0.25 = 75 percent off
    const val GLOW_INTERVAL = 0.05f
    const val GLOW_TIME = 0.5f
    const val GLOW_SIZE = 25f
    const val GLOW_SIZE_INTERVAL = 30f
  }



  override fun apply(stats: MutableShipStatsAPI?, id: String, state: ShipSystemStatsScript.State, effectLevel: Float) {
    //复制粘贴这行
    val ship = (stats?.entity?: return)as ShipAPI

    //创造刷新球特效
    if (!didVisual) {
      addEffect(RefresherOrb(ship))
      didVisual = true
    }

    //effectLevel == 1f 时运行一次
    if (effectLevel < 1f) return

    //回满总装填率
    if(ship.customData.containsKey(aEP_MissilePlatform.ID)){
      val loadingClass = ship.customData[aEP_MissilePlatform.ID] as aEP_MissilePlatform.LoadingMap
      loadingClass.currRate = loadingClass.maxRate
    }

    //为每个武器刷出烟雾
    for (w in ship.allWeapons) {
      if (w.type == WeaponAPI.WeaponType.MISSILE && w.slot.weaponType == WeaponAPI.WeaponType.MISSILE) {
        for (i in 0 until 5) {
          //add cloud
          val colorMult = (Math.random() * 40f).toInt() + 40
          val sizeMult = Math.random().toFloat()
          val ms = aEP_MovingSmoke(w.location)
          ms.setInitVel(speed2Velocity(MathUtils.getRandomNumberInRange(0f, 360f), 1.2f))
          ms.lifeTime = 3f
          ms.fadeIn = 0.1f
          ms.fadeOut = 0.5f
          ms.size = 20 + 40 * sizeMult
          ms.color = Color(colorMult, colorMult, colorMult, (Math.random() * 40).toInt() + 200)
          addEffect(ms)
        }
      }
    }


  }

  override fun unapply(stats: MutableShipStatsAPI, id: String) {
    didVisual = false
  }


  class RefresherOrb internal constructor(var target: CombatEntityAPI) : aEP_BaseCombatEffect() {
    var length: Float
    var advanceDist: Float
    var glowTimer = GLOW_INTERVAL
    override fun advance(amount: Float) {
      super.advance(amount)
      if (length > target.collisionRadius * 3f) {
        cleanup()
      }
      glowTimer = glowTimer + amount
      if (glowTimer > GLOW_INTERVAL) {
        glowTimer = glowTimer - GLOW_INTERVAL
        var xLength = length
        var yLength = target.collisionRadius
        while (yLength >= -target.collisionRadius) {
          val toSpawn = Vector2f(target.location.x + xLength, target.location.y + yLength)
          if (MathUtils.getDistance(toSpawn, target.location) <= target.collisionRadius) {
            var alpha = ((1 - MathUtils.getDistance(toSpawn, target.location) / target.collisionRadius) * 250).toInt()
            alpha = MathUtils.clamp(alpha, 0, 250)
            Global.getCombatEngine().addSmoothParticle(
              toSpawn,
              aEP_ID.VECTOR2F_ZERO,
              GLOW_SIZE, 1f,
              GLOW_TIME,
              Color(0, 250, 0, alpha)
            )
            Global.getCombatEngine().addSmoothParticle(
              toSpawn,
              aEP_ID.VECTOR2F_ZERO,
              GLOW_SIZE, 1f,
              GLOW_TIME,
              Color(0, 250, 0, alpha)
            )
          }
          yLength -= advanceDist
          xLength -= advanceDist
        }
        length += advanceDist
      }
      super.advanceImpl(amount)
    }

    init {
      length = -target.collisionRadius
      val num = (target.collisionRadius * 2f / GLOW_SIZE_INTERVAL).toInt()
      advanceDist = GLOW_SIZE_INTERVAL + target.collisionRadius * 2f % GLOW_SIZE_INTERVAL / num
    }
  }

}